### DISCUSSION TITLE
Input string to output number while preserve the schema

### DISCUSSION THREAD

**Discussion by sumerokr** [open-circle/valibot#115] [Q&A]
Hi,
How can I receive a `string` input, coerce it to `number`, only then validate it as a `number` and also make `Input` and `Output` type correct?

Or, in other words, how can I create a schema that
- returns `string` for `Input<typeof MySchema>`
- coerces incoming `string` value to `number` type before validation
- validates coerced value as a `number`
- returns `number` value
- returns `number` for `Output<typeof MySchema>`
---

What I have found in docs
```
const NumberSchema = coerce(number(), Number);
const numberOutput = parse(NumberSchema, '1234'); // 1234
```
While string can be passed here, its `Input<typeof NumberSchema>` returns `number`.

---

```
transform(string(), Number);
```
Sets correct `Input` (`string`) and `Output` (`number`) types, but validation happens against `string`, before the transformation takes place.

**Comment by fabian-hiller** [comment]
I am aware of this problem, but have not had time to investigate it in detail. Currently, the behavior you expect is not possible. Do you know how it is with other schema libraries?

**Comment by sumerokr** [comment]
Hey, thanks for the answer! Unfortunatelly, I am not aware of others implementation.

**Reply by fabian-hiller** [reply]
Ok. I hope to find time to look into this in 1 to 2 weeks.

**Comment by sillvva** [comment]
@sumerokr Here's a custom transform function, which takes an input type, an output schema, performs the transform, and validates only the output.

```typescript
export function ioTransform<TInput, TOutput extends BaseSchema>(
	outputSchema: TOutput,
	action: (value: TInput) => Output<TOutput>
): BaseSchema<TInput, Output<TOutput>> {
	return {
		...outputSchema,
		_parse(input: TInput, info?: ParseInfo) {
			return outputSchema._parse(action(input), info);
		}
	};
}
```

It does require a custom validator to handle `NaN` values, which do pass the base `number()` check which is `typeof input === "number"`.

![CleanShot 2023-08-28 at 12 41 10@2x](https://github.com/fabian-hiller/valibot/assets/5138384/32b9a943-7c18-4252-930d-55c3e7395d0e)

**Reply by fabian-hiller** [reply]
This looks almost the same as our `coerce` method. The only difference is that the input type can be defined. However, we could improve this. Feel free to create an issue or PR.

**Comment by huyhoang160593** [comment]
> @sumerokr Here's a custom transform function, which takes an input type, an output schema, performs the transform, and validates only the output.
> 
> ```ts
> export function ioTransform<TInput, TOutput extends BaseSchema>(
> 	outputSchema: TOutput,
> 	action: (value: TInput) => Output<TOutput>
> ): BaseSchema<TInput, Output<TOutput>> {
> 	return {
> 		...outputSchema,
> 		_parse(input: TInput) {
> 			const result = outputSchema._parse(action(input));
> 			return result.issues ? result : { output: result.output };
> 		}
> 	};
> }
> ```
> 
> It does require a custom validator to handle `NaN` values, which do pass the base `number()` check which is `typeof input === "number"`.
> 
> ![CleanShot 2023-08-28 at 12 41 10@2x](https://user-images.githubusercontent.com/5138384/263779884-32b9a943-7c18-4252-930d-55c3e7395d0e.png)



> @sumerokr Here's a custom transform function, which takes an input type, an output schema, performs the transform, and validates only the output.
> 
> ```ts
> export function ioTransform<TInput, TOutput extends BaseSchema>(
> 	outputSchema: TOutput,
> 	action: (value: TInput) => Output<TOutput>
> ): BaseSchema<TInput, Output<TOutput>> {
> 	return {
> 		...outputSchema,
> 		_parse(input: TInput) {
> 			const result = outputSchema._parse(action(input));
> 			return result.issues ? result : { output: result.output };
> 		}
> 	};
> }
> ```
> 
> It does require a custom validator to handle `NaN` values, which do pass the base `number()` check which is `typeof input === "number"`.
> 
> ![CleanShot 2023-08-28 at 12 41 10@2x](https://user-images.githubusercontent.com/5138384/263779884-32b9a943-7c18-4252-930d-55c3e7395d0e.png)

Can we add this to the core library, I'm in very need of this to perform my validation to transform string to array string and validate that array

**Reply by fabian-hiller** [reply]
This looks almost the same as our `coerce` method. The only difference is that the input type can be defined. However, we could improve this. Feel free to create an issue or PR.
